sycl/include/sycl/reduction.hpp

```diff
@@ -1082,29 +1082,51 @@ struct NDRangeReduction<
   }
 };
 
-template <typename LocalRedsTy, typename BinOpTy, typename BarrierTy,
-          typename IdentityTy>
-void doTreeReduction(size_t WGSize, size_t LID, bool DisableExtraElem,
-                     IdentityTy Identity, LocalRedsTy &LocalReds, BinOpTy &BOp,
-                     BarrierTy Barrier) {
-  // For work-groups, which size is not power of two, local accessors have
-  // an additional element with index WGSize that is used by the
-  // tree-reduction algorithm. Initialize those additional elements with
-  // identity values here.
-  if (!DisableExtraElem)
-    LocalReds[WGSize] = Identity;
+/// Computes the greatest power-of-two less than or equal to N.
+static inline size_t GreatestPowerOfTwo(size_t N) {
+  if (N == 0)
+    return 0;
+
+  size_t Ret = 1;
+  while ((N >>= 1) != 0)
+    Ret <<= 1;
+  return Ret;
+}
+
+template <typename BarrierTy, typename FuncTy>
+void doTreeReductionHelper(size_t WorkSize, size_t LID, BarrierTy Barrier,
+                           FuncTy Func) {
   Barrier();
-  size_t PrevStep = WGSize;
-  for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
-    if (LID < CurStep)
-      LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
-    else if (!DisableExtraElem && LID == CurStep && (PrevStep & 0x1))
-      LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+
+  // Initial pivot is the greatest power-of-two value smaller or equal to the
+  // work size.
+  size_t Pivot = GreatestPowerOfTwo(WorkSize);
```

Comment:
Thinking out loud - it would be better to have a higher-level interface that would compute that on host and pass to kernels as an argument, I think.